## 1. Counting missing values
<p>Sports clothing and athleisure attire is a huge industry, worth approximately <a href="https://www.statista.com/statistics/254489/total-revenue-of-the-global-sports-apparel-market/">$193 billion in 2021</a> with a strong growth forecast over the next decade! </p>
<p>In this notebook, we play the role of a product analyst for an online sports clothing company. The company is specifically interested in how it can improve revenue. We will dive into product data such as pricing, reviews, descriptions, and ratings, as well as revenue and website traffic, to produce recommendations for its marketing and sales teams.  </p>
<p>The database provided to us, <code>sports</code>, contains five tables, with <code>product_id</code> being the primary key for all of them: </p>
<h3 id="info"><code>info</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_name</code></td>
<td><code>varchar</code></td>
<td>Name of the product</td>
</tr>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>description</code></td>
<td><code>varchar</code></td>
<td>Description of the product</td>
</tr>
</tbody>
</table>
<h3 id="finance"><code>finance</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>listing_price</code></td>
<td><code>float</code></td>
<td>Listing price for product</td>
</tr>
<tr>
<td><code>sale_price</code></td>
<td><code>float</code></td>
<td>Price of the product when on sale</td>
</tr>
<tr>
<td><code>discount</code></td>
<td><code>float</code></td>
<td>Discount, as a decimal, applied to the sale price</td>
</tr>
<tr>
<td><code>revenue</code></td>
<td><code>float</code></td>
<td>Amount of revenue generated by each product, in US dollars</td>
</tr>
</tbody>
</table>
<h3 id="reviews"><code>reviews</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_name</code></td>
<td><code>varchar</code></td>
<td>Name of the product</td>
</tr>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>rating</code></td>
<td><code>float</code></td>
<td>Product rating, scored from <code>1.0</code> to <code>5.0</code></td>
</tr>
<tr>
<td><code>reviews</code></td>
<td><code>float</code></td>
<td>Number of reviews for the product</td>
</tr>
</tbody>
</table>
<h3 id="traffic"><code>traffic</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>last_visited</code></td>
<td><code>timestamp</code></td>
<td>Date and time the product was last viewed on the website</td>
</tr>
</tbody>
</table>
<h3 id="brands"><code>brands</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>brand</code></td>
<td><code>varchar</code></td>
<td>Brand of the product</td>
</tr>
</tbody>
</table>
<p>We will be dealing with missing data as well as numeric, string, and timestamp data types to draw insights about the products in the online store. Let's start by finding out how complete the data is.</p>

In [75]:
%%sql
postgresql:///sports

SELECT
    COUNT(*) AS total_rows,
    COUNT(i.description) AS count_description,
    COUNT(f.listing_price) AS count_listing_price, 
    COUNT(t.last_visited) AS count_last_visited
FROM
    info as i
INNER JOIN finance AS f
    ON i.product_id = f.product_id
INNER JOIN traffic AS t
    ON t.product_id = f.product_id

1 rows affected.


[(3179, 3117, 3120, 2928)]

## 2. Nike vs Adidas pricing
<p>We can see the database contains 3,179 products in total. Of the columns we previewed, only one &mdash; <code>last_visited</code> &mdash; is missing more than five percent of its values. Now let's turn our attention to pricing. </p>
<p>How do the price points of Nike and Adidas products differ? Answering this question can help us build a picture of the company's stock range and customer market. We will run a query to produce a distribution of the <code>listing_price</code> and the count for each price, grouped by <code>brand</code>. </p>

In [77]:
%%sql

SELECT
    brand,
    CAST(listing_price AS int),
    COUNT(f.product_id)
FROM
    brands b
JOIN
    finance f
    ON  b.product_id = f.product_id
GROUP BY brand, listing_price
HAVING listing_price > 0
ORDER BY brand, listing_price DESC


 * postgresql:///sports
77 rows affected.


[('Adidas', 300, 2),
 ('Adidas', 280, 4),
 ('Adidas', 240, 5),
 ('Adidas', 230, 8),
 ('Adidas', 220, 11),
 ('Adidas', 200, 8),
 ('Adidas', 190, 7),
 ('Adidas', 180, 34),
 ('Adidas', 170, 27),
 ('Adidas', 160, 28),
 ('Adidas', 150, 41),
 ('Adidas', 140, 36),
 ('Adidas', 130, 96),
 ('Adidas', 120, 115),
 ('Adidas', 110, 91),
 ('Adidas', 100, 72),
 ('Adidas', 96, 2),
 ('Adidas', 90, 89),
 ('Adidas', 86, 7),
 ('Adidas', 85, 1),
 ('Adidas', 80, 322),
 ('Adidas', 76, 149),
 ('Adidas', 75, 1),
 ('Adidas', 70, 87),
 ('Adidas', 66, 102),
 ('Adidas', 63, 1),
 ('Adidas', 60, 211),
 ('Adidas', 56, 174),
 ('Adidas', 55, 2),
 ('Adidas', 53, 43),
 ('Adidas', 50, 183),
 ('Adidas', 48, 42),
 ('Adidas', 46, 163),
 ('Adidas', 45, 1),
 ('Adidas', 43, 51),
 ('Adidas', 40, 81),
 ('Adidas', 38, 24),
 ('Adidas', 36, 25),
 ('Adidas', 33, 24),
 ('Adidas', 30, 37),
 ('Adidas', 28, 38),
 ('Adidas', 27, 18),
 ('Adidas', 25, 28),
 ('Adidas', 23, 1),
 ('Adidas', 20, 8),
 ('Adidas', 18, 4),
 ('Adidas', 16, 4),
 ('Adidas', 15, 27),
 ('Adidas', 13, 27),
 ('Adidas', 12, 1),
 ('Adidas', 10, 11),
 ('Adidas', 9, 1),
 ('Nike', 200, 1),
 ('Nike', 190, 2),
 ('Nike', 180, 4),
 ('Nike', 170, 14),
 ('Nike', 160, 31),
 ('Nike', 150, 6),
 ('Nike', 140, 12),
 ('Nike', 130, 12),
 ('Nike', 120, 16),
 ('Nike', 110, 17),
 ('Nike', 100, 14),
 ('Nike', 95, 1),
 ('Nike', 90, 13),
 ('Nike', 85, 5),
 ('Nike', 80, 16),
 ('Nike', 79, 1),
 ('Nike', 75, 7),
 ('Nike', 70, 4),
 ('Nike', 65, 1),
 ('Nike', 60, 2),
 ('Nike', 50, 5),
 ('Nike', 48, 1),
 ('Nike', 45, 3),
 ('Nike', 40, 1),
 ('Nike', 30, 2)]

In [78]:
%%sql
SELECT
    COUNT(DISTINCT listing_price) AS unique_prices
FROM
    finance

 * postgresql:///sports
1 rows affected.


[(78,)]

## 3. Labeling price ranges
<p>It turns out there are 77 unique prices for the products in our database, which makes the output of our last query quite difficult to analyze. </p>
<p>Let's build on our previous query by assigning labels to different price ranges, grouping by <code>brand</code> and <code>label</code>. We will also include the total <code>revenue</code> for each price range and <code>brand</code>. </p>

In [80]:
%%sql

SELECT
    b.brand,
    COUNT(f.product_id),
    SUM(revenue) AS total_revenue,
    CASE 
        WHEN listing_price < 42 THEN 'Budget'
        WHEN listing_price >= 42 AND listing_price < 74 THEN 'Average'
        WHEN listing_price >= 74 AND listing_price < 129 THEN 'Expensive'
        ELSE 'Elite' END AS price_category
FROM
    brands b
INNER JOIN
    finance f
    ON  b.product_id = f.product_id
GROUP BY b.brand, price_category
HAVING brand IS NOT NULL
ORDER BY total_revenue DESC

-- Select the brand, a count of all products in the finance table, and total revenue
-- Create four labels for products based on their price range, aliasing as price_category
-- Join brands to finance on product_id
-- Group results by brand and price_category, sort by total_revenue and filter out products missing a value for brand


 * postgresql:///sports
8 rows affected.


[('Adidas', 849, 4626980.069999999, 'Expensive'),
 ('Adidas', 1060, 3233661.060000001, 'Average'),
 ('Adidas', 307, 3014316.8299999987, 'Elite'),
 ('Adidas', 359, 651661.1200000002, 'Budget'),
 ('Nike', 357, 595341.0199999992, 'Budget'),
 ('Nike', 82, 128475.59000000003, 'Elite'),
 ('Nike', 90, 71843.15000000004, 'Expensive'),
 ('Nike', 16, 6623.5, 'Average')]

## 4. Average discount by brand
<p>Interestingly, grouping products by brand and price range allows us to see that Adidas items generate more total revenue regardless of price category! Specifically, <code>"Elite"</code> Adidas products priced \$129 or more typically generate the highest revenue, so the company can potentially increase revenue by shifting their stock to have a larger proportion of these products!</p>
<p>Note we have been looking at <code>listing_price</code> so far. The <code>listing_price</code> may not be the price that the product is ultimately sold for. To understand <code>revenue</code> better, let's take a look at the <code>discount</code>, which is the percent reduction in the <code>listing_price</code> when the product is actually sold. We would like to know whether there is a difference in the amount of <code>discount</code> offered between brands, as this could be influencing <code>revenue</code>.</p>

In [82]:
%%sql

SELECT
    brand,
    100*AVG(discount) as avg_discount
FROM
    brands b
INNER JOIN
    finance f
    ON b.product_id = f.product_id
GROUP BY brand
HAVING brand IS NOT NULL
ORDER BY avg_discount DESC;

 * postgresql:///sports
2 rows affected.


[('Adidas', 33.452427184465606), ('Nike', 0.0)]

## 5. Correlation between revenue and reviews
<p>Strangely, no <code>discount</code> is offered on Nike products! In comparison, not only do Adidas products generate the most revenue, but these products are also heavily discounted! </p>
<p>To improve revenue further, the company could try to reduce the amount of discount offered on Adidas products, and monitor sales volume to see if it remains stable. Alternatively, it could try offering a small discount on Nike products. This would reduce average revenue for these products, but may increase revenue overall if there is an increase in the volume of Nike products sold. </p>
<p>Now explore whether relationships exist between the columns in our database. We will check the strength and direction of a correlation between <code>revenue</code> and <code>reviews</code>. </p>

In [84]:
%%sql

-- Calculate the correlation between reviews and revenue as review_revenue_corr
-- Join the reviews and finance tables on product_id
SELECT
    corr(f.revenue, r.reviews) AS review_revenue_corr
FROM
    reviews r
INNER JOIN
    finance f
    ON r.product_id = f.product_id

 * postgresql:///sports
1 rows affected.


[(0.6518512283481301,)]

## 6. Ratings and reviews by product description length
<p>Interestingly, there is a strong positive correlation between <code>revenue</code> and <code>reviews</code>. This means, potentially, if we can get more reviews on the company's website, it may increase sales of those items with a larger number of reviews. </p>
<p>Perhaps the length of a product's <code>description</code> might influence a product's <code>rating</code> and <code>reviews</code> &mdash; if so, the company can produce content guidelines for listing products on their website and test if this influences <code>revenue</code>. Let's check this out!</p>

In [86]:
%%sql
SELECT
    TRUNC(length(i.description), -2) AS description_length,
    ROUND(CAST(AVG(rating) AS int), 2) AS avg_rating
FROM
    info i
INNER JOIN
    reviews r
    ON  i.product_id = r.product_id
WHERE i.description IS NOT NULL 
GROUP BY description_length
ORDER BY description_length;

 * postgresql:///sports
7 rows affected.


[(Decimal('0'), Decimal('2.00')),
 (Decimal('100'), Decimal('3.00')),
 (Decimal('200'), Decimal('3.00')),
 (Decimal('300'), Decimal('3.00')),
 (Decimal('400'), Decimal('3.00')),
 (Decimal('500'), Decimal('3.00')),
 (Decimal('600'), Decimal('4.00'))]

## 7. Reviews by month and brand
<p>Unfortunately, there doesn't appear to be a clear pattern between the length of a product's <code>description</code> and its <code>rating</code>.</p>
<p>As we know a correlation exists between <code>reviews</code> and <code>revenue</code>, one approach the company could take is to run experiments with different sales processes encouraging more reviews from customers about their purchases, such as by offering a small discount on future purchases. </p>
<p>Let's take a look at the volume of <code>reviews</code> by month to see if there are any trends or gaps we can look to exploit.</p>

In [88]:
%%sql

SELECT
    brand,
    DATE_PART('month', last_visited) AS month,
    COUNT(r.product_id) AS nbr_reviews
FROM
    traffic t
INNER JOIN
    reviews r
    ON t.product_id = r.product_id
INNER JOIN
    brands b
    ON t.product_id = b.product_id
GROUP BY brand, month
HAVING brand IS NOT NULL AND DATE_PART('month', last_visited) IS NOT NULL
ORDER BY brand, month;


 * postgresql:///sports
24 rows affected.


[('Adidas', 1.0, 253),
 ('Adidas', 2.0, 272),
 ('Adidas', 3.0, 269),
 ('Adidas', 4.0, 180),
 ('Adidas', 5.0, 172),
 ('Adidas', 6.0, 159),
 ('Adidas', 7.0, 170),
 ('Adidas', 8.0, 189),
 ('Adidas', 9.0, 181),
 ('Adidas', 10.0, 192),
 ('Adidas', 11.0, 150),
 ('Adidas', 12.0, 190),
 ('Nike', 1.0, 52),
 ('Nike', 2.0, 52),
 ('Nike', 3.0, 55),
 ('Nike', 4.0, 42),
 ('Nike', 5.0, 41),
 ('Nike', 6.0, 43),
 ('Nike', 7.0, 37),
 ('Nike', 8.0, 29),
 ('Nike', 9.0, 28),
 ('Nike', 10.0, 47),
 ('Nike', 11.0, 38),
 ('Nike', 12.0, 35)]

## 8. Footwear product performance
<p>Looks like product reviews are highest in the first quarter of the calendar year, so there is scope to run experiments aiming to increase the volume of reviews in the other nine months!</p>
<p>So far, we have been primarily analyzing Adidas vs Nike products. Now, let's switch our attention to the type of products being sold. As there are no labels for product type, we will create a Common Table Expression (CTE) that filters <code>description</code> for keywords, then use the results to find out how much of the company's stock consists of footwear products and the median <code>revenue</code> generated by these items.</p>

In [90]:
%%sql

WITH cte_footwear AS(
    SELECT
        i.description,
        f.revenue
    FROM
        info i
    INNER JOIN
        finance f
        ON  i.product_id = f.product_id
    WHERE i.description ILIKE '%shoe%'
        OR i.description ILIKE '%trainer%'
        OR i.description ILIKE '%foot%' 
    AND i.description IS NOT NULL
)

SELECT
    COUNT(*) AS nbr_footwear,
    PERCENTILE_DISC(0.5) WITHIN GROUP (ORDER BY revenue) AS median_footwear_revenue
FROM
    cte_footwear;


 * postgresql:///sports
1 rows affected.


[(2700, 3118.36)]

## 9. Clothing product performance
<p>Recall from the first task that we found there are 3,117 products without missing values for <code>description</code>. Of those, 2,700 are footwear products, which accounts for around 85% of the company's stock. They also generate a median revenue of over $3000 dollars!</p>
<p>This is interesting, but we have no point of reference for whether footwear's <code>median_revenue</code> is good or bad compared to other products. So, for our final task, let's examine how this differs to clothing products. We will re-use <code>footwear</code>, adding a filter afterward to count the number of products and <code>median_revenue</code> of products that are not in <code>footwear</code>.</p>

In [92]:
%%sql
WITH cte_footwear AS(
    SELECT
        i.description,
        f.revenue
    FROM
        info i
    INNER JOIN
        finance f
        ON  i.product_id = f.product_id
    WHERE i.description ILIKE '%shoe%'
        OR i.description ILIKE '%trainer%'
        OR i.description ILIKE '%foot%' 
    AND i.description IS NOT NULL
)

SELECT
    COUNT(i.*) AS nbr_clothing_products,
    percentile_disc(0.5) WITHIN GROUP (ORDER BY f.revenue) AS median_clothing_revenue
FROM 
    info i
INNER JOIN 
    finance f 
    ON i.product_id = f.product_id
WHERE i.description NOT IN (SELECT description FROM cte_footwear);


 * postgresql:///sports
1 rows affected.


[(417, 503.82)]